### PR TITLE
Implement RPC Messaging Between Validators

### LINKED ISSUE #452: Implement P2P Message Interactions Between Validator
Hi all,

As part of Eth 2.0 we require validators to communicate with each other on their assigned shard in order to exchange information about attestations. Specifically, we want proposers to receive attestation fields from attesters in a certain shard in order to package that data up into a block to be sent to a beacon node for processing. Workflow is as follows:

1. Proposer packages up basic info into a tentative block
2. Proposer broadcasts the tentative block to other validators in the shard (to attesters).
3. Attesters vote on the tentative block, then they broadcast their attestations to the proposer that sent them the tentative block.
4. Proposer packages those attestations into a request to be sent to a full beacon node via gRPC which will then be transformed into a beacon block by the node.

For this, we need to implement proto p2p messages and interactions between validators.

### PR DESCRIPTION
**Pull request by nisdas**
This resolves #452. This will implement RPC messaging between attesters and proposers, so that proposers can broadcast their potential blocks and attesters can broadcast their attestations to the other validators. 
Requirements for Merge : 
- [x] Proposers are able to broadcast every tentative block to each attester
- [x] Attesters receive the block then broadcast their attestation. 
- [x] Add tests for testing interactions between all the different validators

### CODE REVIEW

**Comment by terenc3t** [comment]
@nisdas 
Don't have to worry about Attesters attesting to the block, that's part of #453 
Also don't have to worry about validators aggregate signatures locally but proposer aggregates it then sends it through grpc for processing to the beacon node. That's out of scope. The important thing is just validators can send attestation and block to each other : )  Thanks a lot for this!

File: validator/attester/service_test.go [hunk 1]
```diff
@@ -41,17 +57,25 @@ func TestAttesterLoop(t *testing.T) {
...
+		BeaconBlock: &pbp2p.BeaconBlock{}}
+
 	att.assignmentChan <- true
+	att.blockBuf <- p2p.Message{Peer: p2p.Peer{}, Data: &shardingp2p.AttestationBroadcast{}}
+	att.blockBuf <- p2p.Message{Peer: p2p.Peer{}, Data: block}
+
 	testutil.AssertLogsContain(t, hook, "Performing attester responsibility")
 	doneChan <- struct{}{}
 	exitRoutine <- true
+	testutil.AssertLogsContain(t, hook, "Attestation Broadcasted to network")
```
Comment (rauljordan) [inline validator/attester/service_test.go]: Attestation broadcasted

File: validator/attester/service.go [hunk 2]
```diff
@@ -22,49 +26,78 @@ type Attester struct {
...
+		case <-a.assignmentChan:
 			log.Info("Performing attester responsibility")
+		case msg := <-a.blockBuf:
+			data, ok := msg.Data.(*shardingp2p.BlockBroadcast)
+			if !ok {
+				log.Error("Received malformed attestation p2p message")
+				continue
+			}
+			attestation := a.CreateAttestation(data.GetBeaconBlock())
+			a.p2p.Broadcast(attestation)
```
Comment (terenc3t) [inline validator/attester/service.go]: This block is not correct. Can you take out this block and refer to #487?

The correct work flow should be:
1. Attester wait for its slot and gets HEAD from beacon node
2. Attester signs HEAD
3. Attester relays HEAD back to beacon node, pass it to rpc service to block chain service to p2p service
4. p2p service will broadcast the attestation

I don't think we should broadcast attestation from client level

File: validator/proposer/service.go [hunk 3]
```diff
@@ -68,10 +88,73 @@ func (p *Proposer) Stop() error {
 	return nil
 }
 
+// SaveBlockToMemory will save the block to memory so that the received
+// attestations can be aggregated in it.
+func (p *Proposer) SaveBlockToMemory(block *pbp2p.BeaconBlock) {
```
Comment (terenc3t) [inline validator/proposer/service.go]: I don't understand why this is done in `proposer` package. This seems more on a beacon node level or validator level scope

File: validator/proposer/service.go [hunk 4]
```diff
@@ -68,10 +88,73 @@ func (p *Proposer) Stop() error {
...
+			exists = true
+			break
+		}
+	}
+	return exists
+}
+
+// AggregateAllSignatures aggregates all the signatures of the attesters. This is currently a
+// stub for now till BLS/other singature schemes are implemented.
+func (p *Proposer) AggregateAllSignatures(attestations []*pbp2p.AttestationRecord) []uint32 {
```
Comment (terenc3t) [inline validator/proposer/service.go]: `AggregateAllSignatures` and `GenerateBitmask` are seems to be out of scope for this PR. This PR is to Implement P2P messaging between validators

File: beacon-chain/simulator/service.go [hunk 5]
```diff
@@ -23,7 +24,7 @@ var log = logrus.WithField("prefix", "simulator")
 type Simulator struct {
 	ctx                    context.Context
 	cancel                 context.CancelFunc
-	p2p                    types.P2P
+	p2p                    shared.P2P
```
Comment (terenc3t) [inline beacon-chain/simulator/service.go]: Why are we moving beacon node p2p to shared?

File: validator/attester/service.go (hunk 2 above)
Comment (nisdas) [reply validator/attester/service.go]: hmm , so what does the proposer actually do then ? 
Why does the attester pass it back to the beacon node then , only for it to be broadcasted again by the beacon node ?
What is the purpose of the beacon node in the above scenario , I was under the impression that attestations and signatures were all aggregated by both attesters and proposers , with the only difference being that the proposer sends the block back to the beacon node through rpc .

File: beacon-chain/simulator/service.go (hunk 5 above)
Comment (nisdas) [reply beacon-chain/simulator/service.go]: that describes the interface for p2p interactions such as `Send`,`Subscribe`, we were only using it in the beacon node before but now with this pr since the validator clients also are using it , it made sense to move it to shared

File: validator/attester/service.go (hunk 2 above)
Comment (terenc3t) [reply validator/attester/service.go]: I think attester should also send the attestation back to the beacon node for book keeping. Unless attester plans to do it himself, but that'd be duplicated effort for attester and proposer. 

Ideally, beacon node should be managing the attestations (ex: DB look up, save old & new attestation in DB, fix oblique parent hashes and find slashing conditions)

File: beacon-chain/simulator/service.go (hunk 5 above)
Comment (terenc3t) [reply beacon-chain/simulator/service.go]: I see. The question is whether we should allow client to have these p2p interactions. That's an open question we should survey with the team.

File: validator/attester/service.go (hunk 2 above)
Comment (nisdas) [reply validator/attester/service.go]: Alright let me try to clear up what the flow of our clients will be like : 
Proposer: 
- In charge of block production, will aggregate all the signature from attesters , and then finally send the block through grpc back to the beacon - node 
Attester:
- Will attest to a potential beacon block, adding in shard id, shard blockhash ,etc and adding the attester bitfield
The issue is how will the the flow would go like, if the attester actually sends the attestation back to the beacon - node to do all this other stuff , ex: fix oblique parent hashes , find slashing conditions, etc and then broadcast it to the proposer, then what is actually the proposer's role in all this then ?
All the work is being done by the beacon-node and attester, the proposer then just aggregates the signatures ? Wouldnt that also be possible to be done by the beacon-node by itself then ?
In the client flow we are describing above, I dont see what role the proposer will have, 
We would also not need to have p2p messaging by validators, the beacon node could simply be connected to the attester clients through rpc.

Comment (terenc3t) [reply validator/attester/service.go]: Not just proposer, attester broadcasts attestations to the rest of the validators. It's every validator's job to aggregate attestations and keep it among themselves. Only proposer gets to propose the block.

Proposer does more in this context. It needs to sign off on its proposed block. A beacon node can not process an incoming beacon block if proposer's signature is not included.

File: shared/p2p/service_test.go [hunk 6]
```diff
@@ -17,13 +17,17 @@ import (
 	swarmt "github.com/libp2p/go-libp2p-swarm/testing"
 	shardpb "github.com/prysmaticlabs/prysm/proto/sharding/p2p/v1"
 	testpb "github.com/prysmaticlabs/prysm/proto/testing"
-	"github.com/prysmaticlabs/prysm/shared"
 	"github.com/sirupsen/logrus"
 	logTest "github.com/sirupsen/logrus/hooks/test"
 )
 
+type testService interface {
```
Comment (prestonvanloon) [inline shared/p2p/service_test.go]: Why do you need this? This is the same interface as shared.Service

File: validator/attester/service.go [hunk 7]
```diff
@@ -22,49 +26,78 @@ type Attester struct {
...
+
 	for {
 		select {
 		case <-done:
 			log.Debug("Attester context closed, exiting goroutine")
 			return
-		case <-p.assignmentChan:
+		case <-a.assignmentChan:
 			log.Info("Performing attester responsibility")
+		case msg := <-a.blockBuf:
```
Comment (prestonvanloon) [inline validator/attester/service.go]: I think this should be in another go routine. If the blockBuf logic is slow then it will end up blocking the assignment logic and vice versa.

I recommend to refactor this method and break out the responsabilities so that it looks something like:
```
func (a *Attester) run(done <- chan struct{}) {
    go a.handleAssignments(done)
    go a.handleBlockBroadcast(done) // Actually, what is this action?
}
```

File: validator/attester/service.go [hunk 8]
```diff
@@ -22,49 +26,78 @@ type Attester struct {
...
+			data, ok := msg.Data.(*shardingp2p.BlockBroadcast)
+			if !ok {
+				log.Error("Received malformed attestation p2p message")
+				continue
+			}
+			attestation := a.CreateAttestation(data.GetBeaconBlock())
+			a.p2p.Broadcast(attestation)
+
+			log.Info("Attestation Broadcasted to network")
```
Comment (prestonvanloon) [inline validator/attester/service.go]: Sentence case please

File: validator/proposer/service.go [hunk 9]
```diff
@@ -68,10 +88,73 @@ func (p *Proposer) Stop() error {
...
+		for i, chunk := range attestation.GetAttesterBitfield() {
+			bitmask[i] = bitmask[i] | chunk
+		}
+	}
+	return bitmask
+}
+
 // run the main event loop that listens for a proposer assignment.
-func (p *Proposer) run(done <-chan struct{}, client pb.ProposerServiceClient) {
+func (p *Proposer) run(delayChan <-chan time.Time, done <-chan struct{}, client pb.ProposerServiceClient) {
```
Comment (prestonvanloon) [inline validator/proposer/service.go]: Likewise feedback with this run method. The amount of responsibility here is increasing signficantly and you have multiple (potentially blocking) logic routines in the single for loop.

Please break this up into smaller pieces, similiar to the feedback I recommended for Attester

File: validator/attester/service.go [hunk 10]
```diff
@@ -22,49 +26,78 @@ type Attester struct {
...
-	defer p.cancel()
+func (a *Attester) Stop() error {
+	defer a.cancel()
 	log.Info("Stopping service")
 	return nil
 }
 
+// CreateAttestation will cause the attester to attest to the provided beacon block.
+func (a *Attester) CreateAttestation(block *pbp2p.BeaconBlock) *pbp2p.AttestationRecord {
+	attestation := &pbp2p.AttestationRecord{}
```
Comment (rawfalafel) [inline validator/attester/service.go]: Does it make sense to use the same `AttestationRecord` data structure here? Here's the fields:

```
type AttestationRecord struct {
	Slot                 uint64 
	ShardId              uint64  
	JustifiedSlot        uint64   
	JustifiedBlockHash   []byte  
	ShardBlockHash       []byte  
	AttesterBitfield     []byte  
	ObliqueParentHashes  [][]byte 
}
```

This is a single attestation, so `AttesterBitfield` doesn't make sense. `ObliqueParentHashes` is important when the proposer includes an attestation for a different fork, but that's obviously unimportant here. Maybe we should create a different data structure for individual attestations.

File: validator/proposer/service.go [hunk 11]
```diff
@@ -111,6 +219,27 @@ func (p *Proposer) run(done <-chan struct{}, client pb.ProposerServiceClient) {
...
+
+		case msg := <-p.attestationBuf:
+			data, ok := msg.Data.(*shardingp2p.AttestationBroadcast)
+			if !ok {
+				log.Error("Received malformed attestation p2p message")
+				continue
+			}
+
+			attestationRecord := data.GetAttestationRecord()
+			block, err := p.GetBlockFromMemory(attestationRecord.GetSlot())
```
Comment (rawfalafel) [inline validator/proposer/service.go]: Looks like this is fetching a block of the same slot as the attestation, but that's not correct. Blocks include attestations for previous slots.

File: shared/p2p/service_test.go (hunk 6 above)
Comment (nisdas) [reply shared/p2p/service_test.go]: I cant use that due to cyclical imports, the go compiler would not be able to compile then . So I just created a similar one here

File: validator/attester/service.go (hunk 7 above)
Comment (nisdas) [reply validator/attester/service.go]: Alright I will break it off for both proposers and attesters into smaller goroutines

**Comment by nisdas** [comment]
@rawfalafel after talking with @terence and how he has implemented #487 , it looks like I might have had the wrong idea of how our clients would work. 

1) The proposer is selected to propose a block at slot N, So they will create a tentative block and send it back through rpc to the beacon-node.
2) The beacon-node then delivers this block to the attester through rpc, the attester attests to the block adding all the relevant fields and sends the attestation for that block back to the beacon-node.
3) The beacon-node processes the attestation, adding oblique parent hash, etc. Then it broadcasts the attestation to the network.
4) The proposer listens to the network for attestations and then adds attestations locally to the block it has. 

In this version the attester does not store the block, it just generates the attestation. The proposer does the aggregation of all the attestations for that block. My thought was that if the beacon-node does not aggregate the attestations, then the only one doing that is the proposer. 

If we agree that attesters/validators have a simple role of simply managing the keys and handling all the shard logic then the proposer is the one aggregating the sigs for the beacon block, when does the proposer pass the block onto the beacon-chain to be processed , after the slot duration(8s) ? How do attestations outside that slot window get delivered to the beacon-node to be aggregated in a previous block, if the proposer has already previously sent that block for processing ? 

If attestations outside the slot window can be added to a beacon-block then wouldnt the beacon-node need to do the aggregating of the sigs ?
Thoughts?

**Comment by rawfalafel** [comment]
Gotcha, I think there's still a misunderstanding on how proposals and attestations work. 

The four steps you outlined are on the right track, but there's a few key points missing. First, there's a 512** second window after a proposal's included in a chain for attestations for that block to be included in the chain. Second, the proposer that includes an attestation for `block N` is different from the proposer that includes `block N` into the chain. 

So the steps should be more like this:

1. A proposer chosen for `slot N` signs a block for that slot and propagates it through the network.
2. Attesters in the committees for `slot N` receive the block, sign it, and propagate these individual attestations through the network.
3. Let's call `slot M` the range of slots between `slot N+1` and `slot N+cycleLength`. Proposers chosen for `slot M` will aggregate individual attestations for `slot N`, along with individual attestations for blocks between `slot M-cycleLength` and `slot M-1`, construct a block for `slot M` with these aggregated attestations, sign the block, and propagate it through the network.

The third step is a little heavy but all its saying is that after a block is proposed, there are `cycleLength` slots for another proposer to include attestations for that block. @terenc3t , is this in line with your understanding?

** 512 = `slotDuration * cycleLength`

**Comment by terenc3t** [comment]
Yup, that's my understanding. Most of those interactions are already/plan to be implemented by proposer and attester responsibility PR. 
This PR is for p2p messaging such as we need to define p2p messages to broadcast block  attestations. Let's leave signature aggregation, save/look up blocks/attestations to the other PRs.

File: beacon-chain/sync/initial-sync/service.go [hunk 12]
```diff
@@ -61,7 +62,7 @@ type SyncService interface {
 type InitialSync struct {
 	ctx                          context.Context
 	cancel                       context.CancelFunc
-	p2p                          types.P2P
+	p2p                          shared.P2P
```
Comment (terenc3t) [inline beacon-chain/sync/initial-sync/service.go]: Should we move p2p back to types since it's only getting used by beacon node?

File: validator/proposer/service.go [hunk 13]
```diff
@@ -68,10 +84,72 @@ func (p *Proposer) Stop() error {
...
+			log.Debug("Proposer context closed, exiting goroutine")
+			return
+		case msg := <-p.attestationBuf:
+			data, ok := msg.Data.(*pbp2p.AttestationBroadcast)
+			if !ok {
+				log.Error("Received malformed attestation p2p message")
+				continue
+			}
+
+			attestationRecord := data.GetAttestationRecord()
```
Comment (terenc3t) [inline validator/proposer/service.go]: Can you explain briefly what attestationRecord and attestationExists are used for?

File: validator/proposer/service.go [hunk 14]
```diff
@@ -95,13 +173,19 @@ func (p *Proposer) run(done <-chan struct{}, client pb.ProposerServiceClient) {
 			}
 			latestBlockHash := blake2b.Sum512(data)
 
+			// To prevent any unaccounted attestations from being added.
+			p.mutex.Lock()
+
+			agSig := p.AggregateAllSignatures(p.pendingAttestation)
+			bitmask := p.GenerateBitmask(p.pendingAttestation)
```
Comment (terenc3t) [inline validator/proposer/service.go]: Slightly out of scope for this PR, this touch bases on proposer responsibility. Need @rauljordan to take a look to make sure there's no duplication of work on his end

File: beacon-chain/sync/initial-sync/service.go (hunk 12 above)
Comment (nisdas) [reply beacon-chain/sync/initial-sync/service.go]: Its actually used by the proposer to subscribe to attestation Broadcasts, so it should stay in the shared package

File: validator/proposer/service.go (hunk 13 above)
Comment (nisdas) [reply validator/proposer/service.go]: attestationRecord is just another name for the attestation by an attester. Attestation exists is to check that there is no duplicated attestation added to the pending attestations by the same attester

File: validator/proposer/service.go (hunk 14 above)
Comment (nisdas) [reply validator/proposer/service.go]: Yeah I removed what I previously implemented for those two functions, so currently those functions are just stubs. Whoever is doing proposer responsibility can just implement it themselves

File: validator/proposer/service.go (hunk 13 above)
Comment (terenc3t) [reply validator/proposer/service.go]: Got it. Thanks!

File: beacon-chain/sync/initial-sync/service.go (hunk 12 above)
Comment (terenc3t) [reply beacon-chain/sync/initial-sync/service.go]: Why not use a feed and pass the attestations from beacon node's rpc to proposer?

The question is should we allow validator to touch p2p?

option 1:
p2p <-> sync service<-> block chain service<->rpc service <-> proposer/attester

option 2:
p2p <-> proposer/attester

Option 2 seems easier, but the down side is it puts more work on the validator package (I think we want to keep it relatively lightweight) and there could be duplicated code between proposer and attester.

Any thoughts? @rauljordan  @rawfalafel  @prestonvanloon

Comment (nisdas) [reply beacon-chain/sync/initial-sync/service.go]: The security concerns are valid but I am not sure if its possible for our validators to only communicate via rpc, I would imagine attesters that would be handling a shard would need to use p2p for a lot of things. Also by only using rpc we reduce our validators to a one is to one relationship with other actors which might make it difficult for us to represent complex relationships between the different actors down the line

File: beacon-chain/params/BUILD.bazel [hunk 15]
```diff
@@ -4,5 +4,8 @@ go_library(
     name = "go_default_library",
     srcs = ["config.go"],
     importpath = "github.com/prysmaticlabs/prysm/beacon-chain/params",
-    visibility = ["//beacon-chain:__subpackages__"],
+    visibility = [
+        "//beacon-chain:__subpackages__",
+        "//validator:__subpackages__",
```
Comment (prestonvanloon) [inline beacon-chain/params/BUILD.bazel]: This shouldn't be allowed. Why are we exposing this to the other project?

File: proto/sharding/p2p/v1/messages.proto [hunk 16]
```diff
@@ -36,4 +36,4 @@ message Signature {
   uint64 v = 1;
   uint64 r = 2;
   uint64 s = 3;
-}
+}
```
Comment (prestonvanloon) [inline proto/sharding/p2p/v1/messages.proto]: Revert this

File: proto/sharding/p2p/v1/messages.pb.go [hunk 17]
```diff
@@ -1,7 +1,7 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
-// source: proto/sharding/p2p/v1/messages.proto
+// source: messages.proto
```
Comment (prestonvanloon) [inline proto/sharding/p2p/v1/messages.pb.go]: Revert these changes

File: beacon-chain/params/BUILD.bazel (hunk 15 above)
Comment (prestonvanloon) [reply beacon-chain/params/BUILD.bazel]: It doesn't look like you are using this, can you revert this change?

File: beacon-chain/sync/initial-sync/service.go (hunk 12 above)
Comment (rawfalafel) [reply beacon-chain/sync/initial-sync/service.go]: > by only using rpc we reduce our validators to a one is to one relationship with other actors

I'm not sure what you mean by this. 

I'm in favor of option 1. The validator is managing private keys; not allowing it to connect directly to the p2p network is a clean abstraction.

Comment (nisdas) [reply beacon-chain/sync/initial-sync/service.go]: I was more referring to the fact that attesters would be validating the state of a shard, so they would be interacting with other actors where rpc might not be the most optimal solution, but in the case of the proposer I will change it to rpc then

File: beacon-chain/params/BUILD.bazel (hunk 15 above)
Comment (nisdas) [reply beacon-chain/params/BUILD.bazel]: Yeah sorry, I forgot to remove this

File: beacon-chain/sync/initial-sync/service.go (hunk 12 above)
Comment (terenc3t) [reply beacon-chain/sync/initial-sync/service.go]: I see. You are referring phase 1 when shard data (w/o execution) comes in to the picture. Even then to interact and play the proof of custody game, attester still needs DB access and it will have to go through with the beacon node via RPC.  If we grant attester DB access, it feels like attester should just become a service in beacon node. Do you guys think attester broadcasts attestation is the right approach here?

File: beacon-chain/rpc/service.go [hunk 18]
```diff
@@ -21,21 +21,23 @@ var log = logrus.WithField("prefix", "rpc")
...
+	announcer             types.CanonicalEventAnnouncer
+	chainService          chainService
+	port                  string
+	listener              net.Listener
+	withCert              string
+	withKey               string
+	grpcServer            *grpc.Server
+	canonicalBlockChan    chan *types.Block
+	canonicalStateChan    chan *types.CrystallizedState
+	proccessedAttestation chan *pbp2p.AttestationRecord
```
Comment (terenc3t) [inline beacon-chain/rpc/service.go]: Is it possible to use `*types.Attestation`? Do we allow validator to import types package from beacon chain?

File: beacon-chain/sync/initial-sync/service.go (hunk 12 above)
Comment (terenc3t) [reply beacon-chain/sync/initial-sync/service.go]: Should we change it back to `types.P2P`?

File: beacon-chain/rpc/service.go (hunk 18 above)
Comment (nisdas) [reply beacon-chain/rpc/service.go]: We would have to expose the types package to the validator, either that or move it to the shared package. I dont think it is necessary to use types.Attestation as eventually when the proposer aggregates the attestations into the block he would have to do using proto.AttestationRecord , in my opinion its best if we keep it in the proto format

File: beacon-chain/sync/initial-sync/service.go (hunk 12 above)
Comment (nisdas) [reply beacon-chain/sync/initial-sync/service.go]: I actually think it would make more sense to keep it here. The main p2p package is in shared folder so it would make sense to keep the interface there also, if you feel strongly about it I will put the interface back in types
